aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::fp32_append_dtype, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    return (*F)(args..., at::kFloat);
+  }
+};
+
+// CastPolicy::promote
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::promote, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    auto to_type = promote_type(at::kHalf, args...);
+    return (*F)(cached_cast(to_type, args)...);
+  }
+};
+
+// CastPolicy::promote_with_tensorlist
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::promote_with_tensorlist, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    auto to_type = promote_type(at::kHalf, args...);
+    // Create vec out of line to ensure its data's lifetime lasts the full duration of (*F)()
+    std::vector<Tensor> vec = get_casted_vector(to_type, args...);
+    return (*F)(cached_cast_with_vector(to_type, vec, args)...);
+  }
+};
+
+/*******************************
+Banned functions
+*******************************/
+
+Tensor binary_cross_entropy_banned(const Tensor &, const Tensor &, const Tensor &, int64_t) {
+  AT_ERROR("torch.nn.functional.binary_cross_entropy and torch.nn.BCELoss are unsafe to autocast.\n"
+           "Many models use a sigmoid layer right before the binary cross entropy layer.\n"
+           "In this case, combine the two layers using torch.nn.functional.binary_cross_entropy_with_logits\n"
+           "or torch.nn.BCEWithLogitsLoss.  binary_cross_entropy_with_logits and BCEWithLogits are\n"
+           "safe to autocast.");
+}
+
+#ifndef USE_STATIC_DISPATCH
+namespace {
+/*****************************************************************************************************************
+This section performs load-time registration for autocast wrappers.
+
+It's debatable at which level operations should be patched.
+
+Option 1:  Patch only at the level of explicit calls into cudnn/cublas (cudnn_convolution, etc),
+because those are the code paths that are guaranteed to use Tensor Cores, therefore they're the ones that
+will benefit most from FP16.  One difficulty with this approach is that convolutions (and other ops) are wrapped
+in several layers of at::* calls.  Several non-explicitly-registered layers (e.g. at::convolution) could be called
+before we reach cudnn_convolution.  Because they are not explicitly registered, these layers would invoke the
+boxed fallback, which would exclude AutocastTensorId, so by the time at::cudnn_convolution is actually
```

Comment:
It's not a boxed fallback anymore; in fact, since the fallthrough doesn't exclude AutocastTensorId, you don't actually have this problem anymore.

Reply:
It's true this comment is out of date, I forgot to update/remove it when I added fallthrough.  Even with fallthrough, though, I want to catch all calls before they hit the autograd history recording, to ensure things get saved for backwards in fp16 for fp16 ops.